cmd/nephe-controller/secret_webhook.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2022 Antrea Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/go-logr/logr"
+	admissionv1 "k8s.io/api/admission/v1"
+	corev1 "k8s.io/api/core/v1"
+	controllerclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	crdv1alpha1 "antrea.io/nephe/apis/crd/v1alpha1"
+)
+
+// SecretValidator is used to validate Secret
+type SecretValidator struct {
+	Client  controllerclient.Client
+	Log     logr.Logger
+	decoder *admission.Decoder
+}
+
+// Handle handles admission requests for Secret.
+func (v *SecretValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	v.Log.Info("Received webhook", "req", req, "operation", req.Operation)
+	switch req.Operation {
+	case admissionv1.Create:
+		return v.validateCreate(req)
+	case admissionv1.Update:
+		return v.validateUpdate(req)
+	case admissionv1.Delete:
+		return v.validateDelete(req)
+	default:
+		v.Log.Info("Invalid admission webhook request")
+	}
+	return admission.Allowed("")
+}
+
+// InjectDecoder injects the decoder.
+func (v *SecretValidator) InjectDecoder(d *admission.Decoder) error { //nolint:unparam
+	v.decoder = d
+	return nil
+}
+
+// allowSecretUpdate returns true only when Secret data key is unchanged
+func (v *SecretValidator) allowSecretUpdate(new *corev1.Secret, old *corev1.Secret, key string) bool {
+	if changed := string(new.Data[key]) != string(old.Data[key]); changed {
```

Comment:
do you really need key to be passed? isnt just compare data field sufficient?

Reply:
As discussed this is required, as Secret Data is a map and it can contain multiple key, value pairs. CPA uses only particular key from this map. So we have to match with key in CPA CR.